### ISSUE TITLE
Issue running Virsorter2_to_anvio parser

### ISSUE THREAD

**Issue by lexikazen** [simroux/VirSorter2_to_Anvio#2]
I tried to run the Virsorter2_to_Anvio parser and I got this error. I'm not really sure if I did something wrong or if there's an issue with my Virsorter2 output that would have caused this? The virsorter_collection.txt and virsorter_additional_info.txt are still generated so there's just a problem with the annotations file. 

`VirSorter2_to_Anvio/virsorter_to_anvio.py -i phageomics/vs2-pass2-combined/ -s splits_basic_info_combined.txt -n all_gene_calls_combined.txt -d /hpc/refdata/virsorter/2.2.3/  -A virsorter_additional_info.txt -C virsorter_collection.txt -F virsorter_annotations.txt`

######################## Printing collection file and additional data file
Traceback (most recent call last):
  File "/scratch/u/al59331/VirSorter2_to_Anvio/virsorter_to_anvio.py", line 254, in <module>
    nb_genes = info_contigs[split_parent][prophage]['nb_genes']
KeyError: 'nb_genes'

**Comment by simroux** [comment]
Hi ! 
Very sorry to only see this now :-( not sure how I missed the notification.

In case you are still looking into this, there seems to be a discrepancy between the format of your Anvi'o split file and what this script expects (specifically this column 'nb_genes'). Could you confirm which version of Anvi'o you were using ?

**Comment by lexikazen** [comment]
Hi, I am using anvi'o 8 to run this.

**Comment by simroux** [comment]
Ok, v8 should work. Can you try the version of the script in the new branch - https://github.com/simroux/VirSorter2_to_Anvio/blob/simroux-patch-1/virsorter_to_anvio.py ? 
This should tell us which sequence specifically is the issue here

**Comment by lexikazen** [comment]
This was the full error message when using the patched version:

c_000000001055 - c_000000001055||full||lt2gene doe not have a nb_genes ?
Traceback (most recent call last):
  File "/scratch/g/jkirby/AD_Metagenomics_Lexi/./VirSorter2_to_Anvio/virsorter_to_anvio.py", line 243, in <module>
    nb_genes = info_contigs[split_parent][virus]['nb_genes']
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: 'nb_genes'

**Comment by simroux** [comment]
Ok, so this looks like an edge case that was not properly handled. The new version at https://github.com/simroux/VirSorter2_to_Anvio/blob/simroux-patch-1/virsorter_to_anvio.py should fix it, but if it doesn't, then could you please share your input files so I can debug on my side ?
thanks !

**Comment by lexikazen** [comment]
I got a new error this time: 

```
Traceback (most recent call last):
  File "/scratch/g/jkirby/AD_Metagenomics_Lexi/./VirSorter2_to_Anvio/virsorter_to_anvio.py", line 115, in <module>
    if vir_tab[1]=="lt2gene" or (len(vir_tab)>=3 & vir_tab[2]=="lt2gene"):
                                               ~~^~~~~~~~~~~~
TypeError: unsupported operand type(s) for &: 'int' and 'str'
```

How should I share my input files? Github won't let me attach them to this comment

**Comment by lexikazen** [comment]
I'm not super familiar with adding things to Github, so I don't know if I did this right, but I invited you as a collaborator on a Repository that includes a .zip folder of the files

**Comment by simroux** [comment]
Ok, I got the files, thanks. Please try this new version: https://github.com/simroux/VirSorter2_to_Anvio/blob/simroux-patch-1/virsorter_to_anvio.py

This should fix the issue, which I believe was coming from the fact that your sequences were run twice through VirSorter 2 (i.e. VirSorter 2 was run on the output of a previous VirSorter 2 run), which was not properly handled.

**Comment by lexikazen** [comment]
Just tried running the newest patch - it ran without error and generated all the necessary files including the annotations.txt, but this time the files were empty? 

If it helps, this was the last few lines of the output after printing all the contig names: 

######################## Printing collection file and additional data file
######################## Now working on the predicted gene functions
Loading gene calls from Anvi'o
Loading phage list
Writing the output function file
0 hallmark genes matched total out of 0

**Comment by simroux** [comment]
Hmm, probably another issue with these "double VirSorter 2" ids. Could you share the two anvi'o files you provide as input (you can add them to the same repo as last time), it should help me understand where the problem is. Thanks !

**Comment by lexikazen** [comment]
I added the input anvi'o files to the files repo. Hopefully that helps!

**Comment by lexikazen** [comment]
Hi, did you ever get a chance to work on this?

**Comment by simroux** [comment]
Sorry, didn't have any time yet. Hopefully in the next few weeks

**Comment by simroux** [comment]
Hi @lexikazen ,
Sorry it took a while. I had a chance to look at the files you shared, and I think the issue is with the way VirSorter2 was run. From what I can understand, it looks like VirSorter2 was run twice on the same contigs after anvi'o loading, which is something this script can not handle. In anvi'o files, contig ids use the format "c_XXX" where XXX is a number (e.g. c_000000000178). In the VirSorter2 output you provided, the contig ids have the following format: "c_XXX||virsorter2_code||virsorter2_code", e.g. "c_000000000178||full_1||full". The script will take these ids, remove the last virsorter2 code, and assumes the rest (i.e. here "c_000000000178||full_1") is the contig id. Since this contig id is not found in Anvi'o files, that will lead to empty output files.

The easiest fix is probably to make sure VirSorter2 is run only once on these contigs, and try the import again ?
